### DISCUSSION TITLE
Case insensitive routes?

### DISCUSSION THREAD

**Discussion by perostlin** [sveltejs/kit#10207] [Q&A] [answered]
Hello,

We're using SvelteKit for our web sites, I wonder if there is a standardized way to make the routes **case insensitive**?

Say I have the following routes structure:
**src/**
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**routes/**
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**blog**

Then if I go to either **/blog**, **/Blog**, **/BLOG**, etc. in my browser I want to end up on the **blog** route.

Any pointers would be appreciated!

**Comment by teemingc** [comment]
You can try using matchers? (I’m not sure if this is the best way)
https://kit.svelte.dev/docs/advanced-routing#matching

The docs show an example of a matcher using a regex. So maybe something a regex with the case-insensitive flag? (I haven’t tested this)

```js
/** @type {import('@sveltejs/kit').ParamMatcher} */
export function match(param) {
    return /^blog$/i.test(param);
}
```

**Reply by perostlin** [reply]
I thought about that too, but the problem is that I want all routes to be case insensitive, so a param matcher would have to be placed on all routes.

But maybe a hook would suffice?

**Reply by teemingc** [reply]
Sorry I misunderstood. Yes, a hook will most certainly suffice!
Maybe something such as below:
```js
import { redirect } from '@sveltejs/kit';

/** @type {import('@sveltejs/kit').Handle} */
export async function handle({ event, resolve }) {
    // redirect to lowercase URL if there's a capital letter in the pathname
	if (event.url.pathname.match(/[A-Z]/)) {
		throw redirect(307, event.url.pathname.toLowerCase());
	}

	const response = await resolve(event);

	return response;
}
```
[StackBlitz](https://stackblitz.com/edit/sveltejs-kit-template-default-s3y6qs?file=src%2Fhooks.server.js)

**Reply by peter-hendy** [reply, marked as answer]
I think it's even better to use the [reroute universal hook](https://kit.svelte.dev/docs/hooks#universal-hooks-reroute)
```typescript
import type { Reroute } from '@sveltejs/kit';

export const reroute: Reroute = ({ url }) => {
	return url.pathname.toLowerCase();
};
```

**Reply by perostlin** [reply]
Late response from my part, sorry about that, both worked fine, the handle makes the url lowercase even in the browser, the reroute keeps the uppercase characters, so for us both works, we will probably go for the reroute just to skip a redirect.

Thanks @eltigerchino and @peter-hendy 🙇‍♂️

**Reply by ambigos1** [reply]
Hi @peter-hendy 

> ```ts
> export const reroute: Reroute = ({ url }) => {
> 	return url.pathname.toLowerCase();
> };
> ```

Hi.
How can I combine url.pathname.toLowerCase(); with the i18n.reroute()?

Tried something like this and still, I am getting 404 in case one letter is in uppercase

This is a code I tried but it didn't solve my issue

/** @type {import('@sveltejs/kit').Reroute} */
export function reroute({ url }) {
	url.pathname = url.pathname.toLowerCase();
	return i18n.reroute()({ url });
}

**Reply by peter-hendy** [reply]
Hey @ambigos1,

While it's hard to know exactly what's going on here, I think the URL is meant to be immutable and so that's probably causing some weird behaviour. I'd try assigning it to a new URL and seeing if that makes a difference:

```javascript 
import type { Reroute } from '@sveltejs/kit';
import { i18n } from '$lib/i18n'; // Adjust import path as needed

export const reroute: Reroute = ({ url }) => {
    const lowercasePath = url.pathname.toLowerCase();
    
    const newUrl = new URL(url);
    newUrl.pathname = lowercasePath;

    return i18n.reroute()({ url: newUrl });
};```
